Feature: CSV Table export

dot-mike opened the pull request:
This adds a new button to the tables for exporting data to CSV using PHP as backend. It has basic feature for downloading the data in the table. Supporting both legacy ajax tables and modern laravel controller based tables.

Filtering is not added yet, but probably will be only supporting controller based tables.

**Tables that has basic CSV export functionality.**
- Device
- Mempools
- Ports
- Processors
- Storage
- Sensors
- Inventory

To add support for a new table, the `table`-tag needs the `data-ajaxurl` tag.
Example `"data-ajaxurl="<?php echo url('/ajax/table/device') ?>">`

New code is added to the backend table controller for enabling this functionality.
- Main export method is `export($Request)` - this is where the the HTTP request will arrive
- SQL Query is prepared by `protected function prepareExportQuery(Request $request)`
- CSV is generated by method `protected function generateCsvResponse($data, $headers, $filename)\`

Helper methods:

- `getExportHeaders()` - Default implementation for headers
- `visibleColumns()` - Gets visible columns for the table
- `formatExportRow($item)` - Default row formatter


For a child class to implemement CSV export functionality, such as DeviceController,   add two new methods:

```
<?php
protected function getExportHeaders()
{
    return [
        'Device ID',
        'Hostname',
        'IP Address',
        // Additional headers...
    ];
}

protected function formatExportRow($device)
{
    return [
        'device_id' => $device->device_id,
        'hostname' => $device->hostname,
        // Additional fields...
    ];
}
```

Screenshot
![bilde](https://github.com/user-attachments/assets/6675473e-6cfd-4856-87b4-34765e25e7de)


DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

murrant commented:
This all seems pretty good. Nice feature!   It would be great if we can minimize the amount of per-table work as much as possible (it already is pretty minimal).

dot-mike commented:
Hi and thanks. I tried to keep it as simple as possible, as you noted. Not sure how else to improve this.

dot-mike commented:
Inventory is left to fix.... and filtering ofc.

murrant commented:
Can we please undo the changes to the legacy code? (vars.inc.php and includes/html/table in particular)
Those need to be ported to the newer style before any changes are made (which should be a separate PR).  Adding more code to legacy code only makes it harder to get them moved to modern code.
Also, don't use $_REQUEST or any other server objects directly, use the Laravel Request object. Hint if you are changing code that references $vars, $_REQUEST, $_GET, or $POST, it is legacy code.

Sorry if I'm a bit harsh, but I would like to get this merged, so we need to pull the scope back a bit. :)

murrant commented on routes/web.php:
```diff
@@ -217,26 +217,32 @@
-            Route::post('vlan-ports', 'VlanPortsController')->name('table.vlan-ports');
-            Route::post('vlan-devices', 'VlanDevicesController')->name('table.vlan-devices');
-            Route::post('vminfo', 'VminfoController');
+            $tableControllers = [
```
Please unroll this loop.

murrant commented on html/ajax_table_export.php:
```diff
@@ -0,0 +1,55 @@
+<?php
```
Don't add this file, we are working to delete these.

dot-mike commented:
Hi and thank for the feedback! I understand legacy is not supposed to be edited except for upgrading to modern code base. But I actually needed this piece of feature myself, so I had no choice to support legacy. I don't have any time at the moment to convert legacy tables, but I can revert the necessary code to support legacy tables.

dot-mike commented:
PR updated. Rebased onto latest changes in Master. Reverted changes to support legacy ajax tables.  Changes tested and functionality works as expected.

I also took the time to fix filtering and add a page limit functionality. By default all export rows is limited to current viewed rows and user must view all rows to export them. I could add another button generate export for all rows without requiring to view all rows? Let me know your thoughts about this one...

@murrant 

Tables that has CSV export functionality.
  - Device
  - Inventory
  - Mempools
  - Outages
  - Ports
  - Processors
  - Sensors
  - Storage

murrant commented:
Nice work!
